File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.java
import java.util.*;

class Solution {
    private static HashMap<String, Integer> mbti = new HashMap<String, Integer>();
    
    public String solution(String[] survey, int[] choices) {
        initMbti();
        calculate(survey, choices);
        return getResult();
    }

    private static void initMbti() {
        mbti.put("RT", 0);
        mbti.put("CF", 0);
        mbti.put("JM", 0);
        mbti.put("AN", 0);
    }
    
    private static void calculate(String[] survey, int[] choices) {
        for(int i=0; i<survey.length; i++) {
            String s = survey[i];
            int choice = choices[i] - 4;
            
            if(!mbti.containsKey(s)) {
                s = new StringBuilder(s).reverse().toString();
                choice = (-1)*choice;
            }
            mbti.put(s, mbti.get(s) + choice);
        }
    }
    
    private static String getResult() {
        System.out.println(mbti);
        StringBuilder result = new StringBuilder();
        
        mbti.forEach((key, value) -> {
            if(value <= 0) result.append(key.charAt(0));
            else result.append(key.charAt(1));
        });
        
        return result.toString();
    }
}

File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.kt
class Solution {
    fun solution(survey: Array<String>, choices: IntArray): String {
        val scoreMap = mutableMapOf<String, Int>(
            "RT" to 0,
            "CF" to 0,
            "JM" to 0,
            "AN" to 0,
        )
        
        survey.zip(choices.map { it - 4 }).forEach {
            if(scoreMap.containsKey(it.first)) {
                scoreMap[it.first] = scoreMap[it.first]!! + it.second
            } else {
                val key = it.first.reversed()
                scoreMap[key] = scoreMap[key]!! - it.second
            }
        }
        
        return scoreMap.map { (key: String, value: Int) ->
            if(value <= 0) key[0] else key[1]
        }.joinToString("")
    }
}